File: app/src/main/java/com/github/vase4kin/teamcityapp/account/create/presenter/CreateAccountPresenterImpl.java
/*
 * Copyright 2016 Andrey Tolpeev
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.vase4kin.teamcityapp.account.create.presenter;

import android.text.TextUtils;

import com.github.vase4kin.teamcityapp.account.create.data.CreateAccountDataManager;
import com.github.vase4kin.teamcityapp.account.create.data.CreateAccountDataModel;
import com.github.vase4kin.teamcityapp.account.create.data.CustomOnLoadingListener;
import com.github.vase4kin.teamcityapp.account.create.data.OnLoadingListener;
import com.github.vase4kin.teamcityapp.account.create.router.CreateAccountRouter;
import com.github.vase4kin.teamcityapp.account.create.tracker.CreateAccountTracker;
import com.github.vase4kin.teamcityapp.account.create.view.CreateAccountView;

import javax.inject.Inject;

/**
 * Impl of {@link CreateAccountPresenter}
 */
public class CreateAccountPresenterImpl implements CreateAccountPresenter, CreateAccountView.ViewListener {

    private final CreateAccountView view;
    private final CreateAccountDataManager dataManager;
    private final CreateAccountDataModel dataModel;
    private final CreateAccountRouter router;
    private final CreateAccountTracker tracker;

    @Inject
    CreateAccountPresenterImpl(CreateAccountView view,
                               CreateAccountDataManager dataManager,
                               CreateAccountDataModel dataModel,
                               CreateAccountRouter router,
                               CreateAccountTracker tracker) {
        this.view = view;
        this.dataManager = dataManager;
        this.dataModel = dataModel;
        this.router = router;
        this.tracker = tracker;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void handleOnCreateView() {
        view.initViews(this);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void handleOnResume() {
        tracker.trackView();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void handleOnDestroy() {
        view.onDestroyView();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void validateUserData(String url, final String userName, final String password, final boolean isSslDisabled) {
        view.hideError();
        if (TextUtils.isEmpty(url)) {
            view.showServerUrlCanNotBeEmptyError();
            return;
        }
        if (TextUtils.isEmpty(userName)) {
            view.showUserNameCanNotBeEmptyError();
            return;
        }
        if (TextUtils.isEmpty(password)) {
            view.showPasswordCanNotBeEmptyError();
            return;
        }
        view.showProgressDialog();
        if (dataModel.hasAccountWithUrl(url, userName)) {
            view.showNewAccountExistErrorMessage();
            view.dismissProgressDialog();
        } else {
            dataManager.authUser(new CustomOnLoadingListener<String>() {
                @Override
                public void onSuccess(String url) {
                    dataManager.saveNewUserAccount(url, userName, password, isSslDisabled, new OnLoadingListener<String>() {
                        @Override
                        public void onSuccess(String serverUrl) {
                            dataManager.initTeamCityService(serverUrl);
                            tracker.trackUserLoginSuccess(!isSslDisabled);
                            view.dismissProgressDialog();
                            view.finish();
                            router.startRootProjectActivityWhenNewAccountIsCreated();
                        }

                        @Override
                        public void onFail(String errorMessage) {
                            view.showCouldNotSaveUserError();
                            view.dismissProgressDialog();
                            tracker.trackUserDataSaveFailed();
                        }
                    });
                }

                @Override
                public void onFail(int code, String errorMessage) {
                    view.showError(errorMessage);
                    view.dismissProgressDialog();
                    tracker.trackUserLoginFailed(errorMessage);
                }
            }, url, userName, password, isSslDisabled);
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void validateGuestUserData(String url, final boolean isSslDisabled) {
        view.hideError();
        if (TextUtils.isEmpty(url)) {
            view.showServerUrlCanNotBeEmptyError();
            return;
        }
        view.showProgressDialog();
        if (dataModel.hasGuestAccountWithUrl(url)) {
            view.showNewAccountExistErrorMessage();
            view.dismissProgressDialog();
        } else {
            dataManager.authGuestUser(new CustomOnLoadingListener<String>() {
                @Override
                public void onSuccess(String url) {
                    dataManager.saveGuestUserAccount(url, isSslDisabled);
                    dataManager.initTeamCityService(url);
                    tracker.trackGuestUserLoginSuccess(!isSslDisabled);
                    view.dismissProgressDialog();
                    view.finish();
                    router.startRootProjectActivityWhenNewAccountIsCreated();
                }

                @Override
                public void onFail(int code, String errorMessage) {
                    view.showError(errorMessage);
                    view.dismissProgressDialog();
                    tracker.trackGuestUserLoginFailed(errorMessage);
                }
            }, url, isSslDisabled);
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void finish() {
        // If guest user account enabled or not
        if (!view.isEmailEmpty()) {
            view.showDiscardDialog();
        } else {
            view.finish();
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void onClick() {
        finish();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void onDisableSslSwitchClick() {
        view.showDisableSslWarningDialog();
    }
}

File: app/src/main/java/com/github/vase4kin/teamcityapp/account/create/presenter/CreateAccountPresenterImpl.kt
/*
 * Copyright 2019 Andrey Tolpeev
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.vase4kin.teamcityapp.account.create.presenter

import com.github.vase4kin.teamcityapp.account.create.data.CreateAccountDataManager
import com.github.vase4kin.teamcityapp.account.create.data.CreateAccountDataModel
import com.github.vase4kin.teamcityapp.account.create.data.CustomOnLoadingListener
import com.github.vase4kin.teamcityapp.account.create.data.OnLoadingListener
import com.github.vase4kin.teamcityapp.account.create.router.CreateAccountRouter
import com.github.vase4kin.teamcityapp.account.create.tracker.CreateAccountTracker
import com.github.vase4kin.teamcityapp.account.create.view.CreateAccountView
import javax.inject.Inject

/**
 * Impl of [CreateAccountPresenter]
 */
class CreateAccountPresenterImpl @Inject constructor(
    private val view: CreateAccountView,
    private val dataManager: CreateAccountDataManager,
    private val dataModel: CreateAccountDataModel,
    private val router: CreateAccountRouter,
    private val tracker: CreateAccountTracker
) : CreateAccountPresenter, CreateAccountView.ViewListener {

    /**
     * {@inheritDoc}
     */
    override fun handleOnCreateView() {
        view.initViews(this)
    }

    /**
     * {@inheritDoc}
     */
    override fun handleOnResume() {
        tracker.trackView()
    }

    /**
     * {@inheritDoc}
     */
    override fun handleOnDestroy() {
        view.onDestroyView()
    }

    /**
     * {@inheritDoc}
     */
    override fun validateUserData(url: String, userName: String, password: String, isSslDisabled: Boolean) {
        view.hideError()
        if (url.isEmpty()) {
            view.showServerUrlCanNotBeEmptyError()
            return
        }
        if (userName.isEmpty()) {
            view.showUserNameCanNotBeEmptyError()
            return
        }
        if (password.isEmpty()) {
            view.showPasswordCanNotBeEmptyError()
            return
        }
        view.showProgressDialog()
        if (dataModel.hasAccountWithUrl(url, userName)) {
            view.showNewAccountExistErrorMessage()
            view.dismissProgressDialog()
        } else {
            dataManager.authUser(object : CustomOnLoadingListener<String> {
                override fun onSuccess(url: String) {
                    dataManager.saveNewUserAccount(
                        url,
                        userName,
                        password,
                        isSslDisabled,
                        object : OnLoadingListener<String> {
                            override fun onSuccess(serverUrl: String) {
                                dataManager.initTeamCityService(serverUrl)
                                tracker.trackUserLoginSuccess(!isSslDisabled)
                                view.dismissProgressDialog()
                                view.finish()
                                router.startRootProjectActivityWhenNewAccountIsCreated()
                            }

                            override fun onFail(errorMessage: String) {
                                view.showCouldNotSaveUserError()
                                view.dismissProgressDialog()
                                tracker.trackUserDataSaveFailed()
                            }
                        })
                }

                override fun onFail(code: Int, errorMessage: String) {
                    view.showError(errorMessage)
                    view.dismissProgressDialog()
                    tracker.trackUserLoginFailed(errorMessage)
                }
            }, url, userName, password, isSslDisabled, false)
        }
    }

    /**
     * {@inheritDoc}
     */
    override fun validateGuestUserData(url: String, isSslDisabled: Boolean) {
        view.hideError()
        if (url.isEmpty()) {
            view.showServerUrlCanNotBeEmptyError()
            return
        }
        view.showProgressDialog()
        if (dataModel.hasGuestAccountWithUrl(url)) {
            view.showNewAccountExistErrorMessage()
            view.dismissProgressDialog()
        } else {
            dataManager.authGuestUser(object : CustomOnLoadingListener<String> {
                override fun onSuccess(url: String) {
                    dataManager.saveGuestUserAccount(url, isSslDisabled)
                    dataManager.initTeamCityService(url)
                    tracker.trackGuestUserLoginSuccess(!isSslDisabled)
                    view.dismissProgressDialog()
                    view.finish()
                    router.startRootProjectActivityWhenNewAccountIsCreated()
                }

                override fun onFail(statusCode: Int, errorMessage: String) {
                    view.showError(errorMessage)
                    view.dismissProgressDialog()
                    tracker.trackGuestUserLoginFailed(errorMessage)
                }
            }, url, isSslDisabled, false)
        }
    }

    /**
     * {@inheritDoc}
     */
    override fun finish() {
        // If guest user account enabled or not
        if (!view.isEmailEmpty) {
            view.showDiscardDialog()
        } else {
            view.finish()
        }
    }

    /**
     * {@inheritDoc}
     */
    override fun onClick() {
        finish()
    }

    /**
     * {@inheritDoc}
     */
    override fun onDisableSslSwitchClick() {
        view.showDisableSslWarningDialog()
    }
}
